testng-core/src/main/java/org/testng/DataProviderHolder.java

```diff
@@ -15,12 +19,32 @@ public class DataProviderHolder {
   private final Map<Class<?>, IDataProviderListener> listeners = Maps.newConcurrentMap();
   private final Collection<IDataProviderInterceptor> interceptors = Sets.newHashSet();
 
+  private final IConfiguration configuration;
+
+  public DataProviderHolder(IConfiguration configuration) {
+    this.configuration = configuration;
+  }
+
   public Collection<IDataProviderListener> getListeners() {
-    return Collections.unmodifiableCollection(listeners.values());
+    List<IDataProviderListener> original = Lists.newArrayList(listeners.values());
+    ListenerComparator comparator = getConfiguration().getListenerComparator();
+    if (comparator != null) {
+      original.sort(comparator::compare);
+    }
+    return Collections.unmodifiableCollection(original);
+  }
+
+  public IConfiguration getConfiguration() {
+    return Objects.requireNonNull(configuration);
   }
 
   public Collection<IDataProviderInterceptor> getInterceptors() {
-    return Collections.unmodifiableCollection(interceptors);
+    List<IDataProviderInterceptor> original = Lists.newArrayList(interceptors);
+    ListenerComparator comparator = getConfiguration().getListenerComparator();
```

Comment:
Avoid duplication and create a private static method that sorts a list when a comparator is not null and return an unmodifiableCollection

Reply:
When i build a method that looks like 

```java
  private Collection<? extends ITestNGListener> sort(Collection<? extends ITestNGListener> initial) {
    ListenerComparator comparator = configuration.getListenerComparator();
    if (comparator == null) {
      return initial;
    }
    List<? extends ITestNGListener> original = Lists.newArrayList(initial);
    original.sort(comparator);
    return Collections.unmodifiableCollection(original);
  }
```

then the caller would look like below

```java
  public Collection<IDataProviderInterceptor> getInterceptors() {
    return (Collection<IDataProviderInterceptor>) sort(interceptors);
  }
```

triggering an `Unchecked cast` edit check on this method. I can add a `@SuppressWarnings`. Is that fine ?